pkg/blockbuilder/blockbuilder_test.go

```diff
@@ -303,6 +486,116 @@ func TestBlockBuilder_ReachHighWatermarkBeforeLastCycleSection(t *testing.T) {
+	kafkaClient := mustKafkaClient(t, kafkaAddr)
+
+	cfg, overrides := blockBuilderPullModeConfig(t, kafkaAddr)
+
+	// Producing backlog of records in partition 0.
+	// For this test, the last-produced record must belong to the second to last cycle's section; i.e. the last section has nothing to consume.
+	cycleEndStartup := cycleEndAtStartup(time.Now(), cfg.ConsumeInterval, cfg.ConsumeIntervalBuffer)
+
+	var producedSamples []mimirpb.Sample
+	kafkaRecTime := cycleEndStartup.Truncate(cfg.ConsumeInterval).Add(-5 * time.Hour).Add(29 * time.Minute)
+	firstKafkaRecTime := kafkaRecTime.Add(-time.Minute)
+	lastKafkaRecTime := cycleEndStartup.Truncate(cfg.ConsumeInterval).Add(-cfg.ConsumeInterval)
+	for kafkaRecTime.Before(lastKafkaRecTime) {
+		samples := produceSamples(ctx, t, kafkaClient, 0, kafkaRecTime, "1", kafkaRecTime.Add(-time.Minute))
+		producedSamples = append(producedSamples, samples...)
+
+		kafkaRecTime = kafkaRecTime.Add(cfg.ConsumeInterval / 2)
+	}
+	require.NotEmpty(t, producedSamples)
+
+	var p1RecTime time.Time
+
+	// Produce extra record to partition 1 to verify that block-builder finished the whole cycle and didn't get stuck consuming partition 0.
+	{
+		const partition = 1
+		kafkaRecTime := cycleEndStartup.Truncate(cfg.ConsumeInterval).Add(-time.Minute)
+		p1RecTime = kafkaRecTime
+		samples := floatSample(kafkaRecTime.UnixMilli(), 100)
+		val := createWriteRequest(t, "1", samples, nil)
+		produceRecords(ctx, t, kafkaClient, kafkaRecTime, "1", testTopic, partition, val)
+		producedSamples = append(producedSamples, samples...)
+	}
+
+	scheduler := &mockSchedulerClient{}
+	scheduler.addJob(
+		schedulerpb.JobKey{
+			Id:    "test-job-p0-4898",
+			Epoch: 90002,
+		},
+		schedulerpb.JobSpec{
+			Topic:          testTopic,
+			Partition:      0,
+			StartOffset:    0,
+			EndOffset:      int64(len(producedSamples)),
+			CommitRecTs:    firstKafkaRecTime,
+			LastSeenOffset: 0,
+			LastBlockEndTs: time.UnixMilli(0),
+			CycleEndTs:     kafkaRecTime,
+			CycleEndOffset: int64(len(producedSamples) - 2),
+		},
+	)
+	// A second job to include the extra record in partition 1.
+	scheduler.addJob(
+		schedulerpb.JobKey{
+			Id:    "test-job-p1-4899",
+			Epoch: 90070,
+		},
+		schedulerpb.JobSpec{
+			Topic:          testTopic,
+			Partition:      1,
+			StartOffset:    0,
+			EndOffset:      1,
+			CommitRecTs:    p1RecTime,
+			LastSeenOffset: 0,
+			LastBlockEndTs: time.UnixMilli(0),
+			CycleEndTs:     p1RecTime.Add(1 * time.Minute),
+			CycleEndOffset: 1,
+		},
+	)
+
+	reg := prometheus.NewPedanticRegistry()
+	bb, err := newWithSchedulerClient(cfg, test.NewTestingLogger(t), reg, overrides, scheduler)
+	require.NoError(t, err)
+
+	require.NoError(t, services.StartAndAwaitRunning(ctx, bb))
+	t.Cleanup(func() {
+		require.NoError(t, services.StopAndAwaitTerminated(ctx, bb))
+	})
+
+	// Wait for both jobs to complete.
+	require.Eventually(t, func() bool {
+		return bb.jobIteration.Load() >= 2
```

Comment:
Shall we also check for completed counts with scheduler.counts()? Should completed jobs be 2?